### PR TITLE
Unbounded Binary Search 

### PR DESCRIPTION
**Pull request by AmanBisht01**
### Have you read the [Contributing Guidelines on Pull Requests](https://github.com/TesseractCoding/NeoAlgo/blob/master/CONTRIBUTING.md#pull-requests)?
Yes

### Description

Find the point where a monotonically increasing function becomes positive first time.

### Checklist

- [x] I've read the contribution guidelines.
- [x] I've checked the issue list before deciding what to submit.
- [x] I've edited the `README.md` and link to my code.

## Related Issues or Pull Requests

Issue-Unbounded Binary Search #2910

### CODE REVIEW

**Review by AmanBisht01 — COMMENTED** [review]
Review Requested

File: C-Plus-Plus/search/Unbounded_Binary_Search.cpp [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+input x. Find the value ‘n’ where f() becomes positive for the first time.
+ Since f() is monotonically increasing, values of f(n+1), f(n+2),… must be positive and 
+ values of f(n-2), f(n-3), .. must be negative.
+*/
+
+// C++ code for binary search 
+#include<bits/stdc++.h> 
+using namespace std; 
+
+int binarySearch(int low, int high); // prototype
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/search/Unbounded_Binary_Search.cpp]: Remove the redundant comments.

File: C-Plus-Plus/search/Unbounded_Binary_Search.cpp [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+// as f(x) = x^2 - 10*x - 20 Note that 
+// f(x) can be any monotonocally increasing function 
+int f(int x) { return (x*x - 10*x - 20); } 
+
+// Returns the value x where above 
+// function f() becomes positive 
+// first time. 
+int findFirstPositive() 
+{ 
+	// When first value itself is positive
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/search/Unbounded_Binary_Search.cpp]: Remove the redundant comments.

File: C-Plus-Plus/search/Unbounded_Binary_Search.cpp [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+int binarySearch(int low, int high); // prototype 
+
+// Let's take an example function 
+// as f(x) = x^2 - 10*x - 20 Note that 
+// f(x) can be any monotonocally increasing function 
+int f(int x) { return (x*x - 10*x - 20); } 
+
+// Returns the value x where above 
+// function f() becomes positive 
+// first time.
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/search/Unbounded_Binary_Search.cpp]: Remove the redundant comments.

File: C-Plus-Plus/search/Unbounded_Binary_Search.cpp [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+// Returns the value x where above 
+// function f() becomes positive 
+// first time. 
+int findFirstPositive() 
+{ 
+	// When first value itself is positive 
+	if (f(0) > 0) 
+		return 0; 
+
+	// Find 'high' for binary search by repeated doubling
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/search/Unbounded_Binary_Search.cpp]: Remove the redundant comments.

File: C-Plus-Plus/search/Unbounded_Binary_Search.cpp [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+	// When first value itself is positive 
+	if (f(0) > 0) 
+		return 0; 
+
+	// Find 'high' for binary search by repeated doubling 
+	int i = 1; 
+	while (f(i) <= 0) 
+		i = i*2; 
+
+	// Call binary search
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/search/Unbounded_Binary_Search.cpp]: Remove the redundant comments.

File: C-Plus-Plus/search/Unbounded_Binary_Search.cpp [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+	int i = 1; 
+	while (f(i) <= 0) 
+		i = i*2; 
+
+	// Call binary search 
+	return binarySearch(i/2, i); 
+} 
+
+// Searches first positive value 
+// of f(i) where low <= i <= high
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/search/Unbounded_Binary_Search.cpp]: Remove the redundant comments.

File: C-Plus-Plus/search/Unbounded_Binary_Search.cpp [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+{ 
+	if (high >= low) 
+	{ 
+		int mid = low + (high - low)/2; /* mid = (low + high)/2 */
+
+		// If f(mid) is greater than 0 and 
+		// one of the following two 
+		// conditions is true: 
+		// a) mid is equal to low 
+		// b) f(mid-1) is negative
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/search/Unbounded_Binary_Search.cpp]: Remove the redundant comments.

File: C-Plus-Plus/search/Unbounded_Binary_Search.cpp [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+
+		// If f(mid) is greater than 0 and 
+		// one of the following two 
+		// conditions is true: 
+		// a) mid is equal to low 
+		// b) f(mid-1) is negative 
+		if (f(mid) > 0 && (mid == low || f(mid-1) <= 0)) 
+			return mid; 
+
+		// If f(mid) is smaller than or equal to 0
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/search/Unbounded_Binary_Search.cpp]: Remove the redundant comments.

File: C-Plus-Plus/search/Unbounded_Binary_Search.cpp [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+		// conditions is true: 
+		// a) mid is equal to low 
+		// b) f(mid-1) is negative 
+		if (f(mid) > 0 && (mid == low || f(mid-1) <= 0)) 
+			return mid; 
+
+		// If f(mid) is smaller than or equal to 0 
+		if (f(mid) <= 0) 
+			return binarySearch((mid + 1), high); 
+		else // f(mid) > 0
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/search/Unbounded_Binary_Search.cpp]: Remove the redundant comments.

File: C-Plus-Plus/search/Unbounded_Binary_Search.cpp [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+
+		// If f(mid) is smaller than or equal to 0 
+		if (f(mid) <= 0) 
+			return binarySearch((mid + 1), high); 
+		else // f(mid) > 0 
+			return binarySearch(low, (mid -1)); 
+	} 
+
+	/* Return -1 if there is no 
+	positive value in given range */
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/search/Unbounded_Binary_Search.cpp]: Remove the redundant comments.

File: C-Plus-Plus/search/Unbounded_Binary_Search.cpp [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+	return -1; 
+} 
+
+/* Driver code */
+int main() 
+{ 
+	cout<<"The value n where f() becomes" << 
+		"positive first is "<< findFirstPositive();
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/search/Unbounded_Binary_Search.cpp]: Add in a single line.

File: C-Plus-Plus/search/Unbounded_Binary_Search.cpp [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+2.O(Logn): We can’t directly apply Binary Search as we don’t have an upper limit or high index.
+The idea is to do repeated doubling until we find a positive value, i.e., check values of f() 
+for following values until f(i) becomes positive. Let 'high' be the value of i when f() becomes 
+positive for first time.
+We can apply Binary Search now, we can use ‘high/2’ as low and ‘high’ as high indexes in binary search.
+The result n must lie between ‘high/2’ and ‘high’.
+
+##Time Conplexity:O(log n)
+##Space Complexity:O(1)
+*/
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/search/Unbounded_Binary_Search.cpp]: - Add sample I/O at the end of the Code

File: C-Plus-Plus/search/Unbounded_Binary_Search.cpp [hunk 13]
```diff
@@ -0,0 +1,95 @@
+/*
+Given a function ‘int f(unsigned int x)’ which takes a non-negative integer ‘x’ 
+as input and returns an integer as output. The function is monotonically increasing 
+with respect to value of x, i.e., the value of f(x+1) is greater than f(x) for every 
+input x. Find the value ‘n’ where f() becomes positive for the first time.
+ Since f() is monotonically increasing, values of f(n+1), f(n+2),… must be positive and 
+ values of f(n-2), f(n-3), .. must be negative.
+*/
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/search/Unbounded_Binary_Search.cpp]: Use Multiline comments

**Review by ankitaggarwal23 — CHANGES_REQUESTED** [review]
Few Changes:

- Use proper names for variables

File: C-Plus-Plus/search/Unbounded_Binary_Search.cpp (hunk 13 above)
Comment (AmanBisht01) [reply C-Plus-Plus/search/Unbounded_Binary_Search.cpp]: done

File: C-Plus-Plus/search/Unbounded_Binary_Search.cpp [hunk 14]
```diff
@@ -0,0 +1,71 @@
+/*
+Given a function ‘int f(unsigned int x)’ which takes a non-negative integer ‘x’ 
+as input and returns an integer as output. The function is monotonically increasing 
+with respect to value of x, i.e., the value of f(x+1) is greater than f(x) for every 
+input x. Find the value ‘n’ where f() becomes positive for the first time.
+Since f() is monotonically increasing, values of f(n+1), f(n+2),… must be positive and 
+values of f(n-2), f(n-3), .. must be negative.
+*/
+
+
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/search/Unbounded_Binary_Search.cpp]: Remove one extra line

File: C-Plus-Plus/search/Unbounded_Binary_Search.cpp [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+#include<bits/stdc++.h> 
+using namespace std; 
+
+int binarySearch(int low, int high); 
+
+// Let's take an example function 
+// f(x) = x^2 - 10*x - 20 
+int f(int x) { return (x*x - 10*x - 20); }
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/search/Unbounded_Binary_Search.cpp]: Rewrite this to: Also give proper function name as Don't use single letter to define function as it looks not meaningful.
int f(int x) 
{ 
       return (x*x - 10*x - 20); 
}

File: C-Plus-Plus/search/Unbounded_Binary_Search.cpp [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+#include<bits/stdc++.h> 
+using namespace std; 
+
+int binarySearch(int low, int high); 
+
+// Let's take an example function 
+// f(x) = x^2 - 10*x - 20 
+int f(int x) { return (x*x - 10*x - 20); } 
+
+
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/search/Unbounded_Binary_Search.cpp:20]: Remove one extra line

File: C-Plus-Plus/search/Unbounded_Binary_Search.cpp [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+int findFirstPositive() 
+{ 
+
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/search/Unbounded_Binary_Search.cpp]: Remove one extra line

File: C-Plus-Plus/search/Unbounded_Binary_Search.cpp [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+int f(int x) { return (x*x - 10*x - 20); } 
+
+
+int findFirstPositive() 
+{ 
+	 
+	if (f(0) > 0) 
+		return 0; 
+
+
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/search/Unbounded_Binary_Search.cpp]: Remove one extra line

File: C-Plus-Plus/search/Unbounded_Binary_Search.cpp [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+	 
+	if (f(0) > 0) 
+		return 0; 
+
+	
+	int i = 1; 
+	while (f(i) <= 0) 
+		i = i*2; 
+
+
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/search/Unbounded_Binary_Search.cpp]: Remove one extra line

File: C-Plus-Plus/search/Unbounded_Binary_Search.cpp [hunk 20]
```diff
@@ -0,0 +1,71 @@
...
+	return binarySearch(i/2, i); 
+} 
+
+
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/search/Unbounded_Binary_Search.cpp]: Remove one extra line

File: C-Plus-Plus/search/Unbounded_Binary_Search.cpp [hunk 21]
```diff
@@ -0,0 +1,71 @@
...
+		if (f(mid) > 0 && (mid == low || f(mid-1) <= 0)) 
+			return mid; 
+
+		if (f(mid) <= 0) 
+			return binarySearch((mid + 1), high); 
+		else 
+			return binarySearch(low, (mid -1)); 
+	} 
+
+
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/search/Unbounded_Binary_Search.cpp]: Remove one extra line

File: C-Plus-Plus/search/Unbounded_Binary_Search.cpp [hunk 22]
```diff
@@ -0,0 +1,71 @@
...
+	cout<<"The value n where f() becomes positive first is "<< findFirstPositive(); 
+	return 0; 
+} 
+
+/*
+Input  : x^2 - 10*x - 20
+Output : The value n where f() becomes positive first is 12
+
+
+
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/search/Unbounded_Binary_Search.cpp]: Remove two extra line

File: C-Plus-Plus/search/Unbounded_Binary_Search.cpp [hunk 23]
```diff
@@ -0,0 +1,71 @@
...
+Time Conplexity:O(log n)
+Space Complexity:O(1)
+*/
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/search/Unbounded_Binary_Search.cpp:63]: Add a new line at EOF

File: C-Plus-Plus/search/Unbounded_Binary_Search.cpp [hunk 24]
```diff
@@ -0,0 +1,71 @@
...
+/* Driver code */
+int main() 
+{ 
+	cout<<"The value n where f() becomes positive first is "<< findFirstPositive(); 
+	return 0; 
+} 
+
+/*
+Input  : x^2 - 10*x - 20
+Output : The value n where f() becomes positive first is 12
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/search/Unbounded_Binary_Search.cpp]: Make it as:
/*
SAMPLE OUTPUT:

Input  : x^2 - 10*x - 20
Output : The value n where f() becomes positive first is 12

**Review by ankitaggarwal23 — CHANGES_REQUESTED** [review]

**Review by ankitaggarwal23 — APPROVED** [review]

**Pull request merged**